Skip unknown opcode blocks while parsing

paulkaplan opened the pull request:
### Resolves

_What Github issue does this resolve (please include link)?_

Fixes https://github.com/LLK/scratch-gui/issues/2392

### Proposed Changes

_Describe what this Pull Request does_

Be very slightly more careful when parsing blocks to make sure they exist before trying to use them.

### Reason for Changes

_Explain why these changes should be made_

Basically we are treating `parseBlock` like it always returns an array, but it can return undefined if the specmap lookup fails. So for now, just skip those blocks in the processing. 

![image](https://user-images.githubusercontent.com/654102/41619483-724364f4-73d4-11e8-8298-89b1b83ee9ba.png)



### Test Coverage

_Please show how you have added tests to cover your changes_

Added an sb2 test fixture derived from the project in https://github.com/LLK/scratch-gui/issues/2392 

The test explains what is in the fixture vs. what the expectations assert

paulkaplan committed:
Skip unknown opcode blocks while parsing

colbygk commented:
You are the **best**

towerofnix commented:
Is there a warning for skipped blocks, so that if, for example, a project relies on a block that wasn't implemented in 3.0 (somehow), we would see that the block is missing in the console?

kchadha commented on src/serialization/sb2.js:
```diff
@@ -133,17 +133,20 @@ const parseBlockList = function (blockList, addBroadcastMsg, getVariableId, exte
-        if (previousBlock) {
-            parsedBlock.parent = previousBlock.id;
-            previousBlock.next = parsedBlock.id;
+        if (parsedBlockAndComments) { // Could fail due to unrecognized op-codes
```
If a block that is undefined in Scratch 3.0 had a block comment on it in Scratch 2.0, it should probably get removed from the comments list. Right now, it gets attached to the next block in the stack (if there are more blocks in the stack) or doesn't display anywhere if there are no more blocks in the stack, but there's still a comment associated with it (that isn't done being parsed -- e.g. a blockId gets associated with it), so saving and loading that project in 3.0 then fails.

kchadha replied:
So that actually means two things: 
1) `commentIndex` should still be updated in `parseBlock` even if it returns undefined (because that commentIndex refers to a position in a 2.0 stack).
2) Any block comment associated with an undefined block should get removed from the target's comments map (so that it doesn't get serialized later)

Testing this made me realized that this is also an actual issue for how I'm handling importing comments from 2.0 where multiple block comments are attached to the same block. Tracking in issue #1257

paulkaplan committed:
Deal with comments attached to undefined blocks

Also fix an issue where you couldn't save/load projects that had multiple comments linked to a single block.

kchadha commented on src/serialization/sb2.js:
```diff
@@ -133,11 +133,12 @@ const parseBlockList = function (blockList, addBroadcastMsg, getVariableId, exte
         // eslint-disable-next-line no-use-before-define
         const parsedBlockAndComments = parseBlock(block, addBroadcastMsg, getVariableId,
             extensions, comments, commentIndex);
+
```
Nit: stray newline

kchadha commented on src/serialization/sb2.js:
```diff
@@ -706,9 +707,19 @@ const specMapBlock = function (block) {
     if (!blockMetadata) {
-        return;
+        // No block opcode found, exclude this block, increment the comment id and
+        // send all linked comments back to zero/zero to prevent serialization issues.
```
You should probably also say that we're making them workspace comments

kchadha reviewed (commented):
Looks good apart from two nits.

paulkaplan committed:
Fix nits

paulkaplan commented:
@kchadha thanks, fixed those two issues. Struggling with build issues right now, but I assume good to merge after that is resolved?

kchadha commented on src/serialization/sb2.js:
```diff
@@ -711,8 +710,9 @@ const parseBlock = function (sb2block, addBroadcastMsg, getVariableId, extension
     if (!blockMetadata) {
-        // No block opcode found, exclude this block, increment the comment id and
-        // send all linked comments back to zero/zero to prevent serialization issues.
+        // No block opcode found, exclude this block, increment the comment id,
```
Last nit, sorry! `comment id` should be `commentIndex` above.

paulkaplan committed:
Fix comment
